client/rpc/src/state/tests.rs

```diff
@@ -556,36 +544,33 @@ fn should_return_runtime_version() {
 		[\"0xf78b278be53f454c\",2],[\"0xab3c0572291feb8b\",1],[\"0xbc9d89904f5b923f\",1]],\
 		\"transactionVersion\":1}";
 
-	let runtime_version = executor::block_on(api.runtime_version(None.into())).unwrap();
+	let runtime_version = api.runtime_version(None.into()).await.unwrap();
 	let serialized = serde_json::to_string(&runtime_version).unwrap();
 	assert_eq!(serialized, result);
 
 	let deserialized: RuntimeVersion = serde_json::from_str(result).unwrap();
 	assert_eq!(deserialized, runtime_version);
 }
 
-#[test]
-fn should_notify_on_runtime_version_initially() {
-	let (subscriber, id, mut transport) = Subscriber::new_test("test");
-
-	{
-		let client = Arc::new(substrate_test_runtime_client::new());
-		let (api, _child) = new_full(
-			client.clone(),
-			SubscriptionManager::new(Arc::new(TaskExecutor)),
-			DenyUnsafe::No,
-			None,
-		);
-
-		api.subscribe_runtime_version(Default::default(), subscriber);
+#[tokio::test]
+async fn should_notify_on_runtime_version_initially() {
+	let client = Arc::new(substrate_test_runtime_client::new());
+	let (api, _child) =
+		new_full(client, SubscriptionTaskExecutor::new(TaskExecutor), DenyUnsafe::No, None);
 
-		// assert id assigned
-		assert!(matches!(executor::block_on(id), Ok(Ok(SubscriptionId::String(_)))));
-	}
+	let api_rpc = api.into_rpc();
+	let (_sub_id, mut sub_rx) =
+		api_rpc.test_subscription("state_subscribeRuntimeVersion", None).await;
 
 	// assert initial version sent.
-	executor::block_on((&mut transport).take(1).collect::<Vec<_>>());
-	assert!(executor::block_on(transport.next()).is_none());
+	assert_matches!(timeout_secs(1, sub_rx.next()).await, Ok(Some(_)));
+
+	// TODO (jsdw): The channel remains open here, so waiting for another message will time out.
+	// Previously the channel returned None.
+	// TODO (dp): I think this is a valid concern; our version swallows the `None` (in the
+	// `take_while` call I guess?). I guess this test does what is says on the tin though: check
+	// that we get the current runtime version when subscribing.
+	assert_matches!(timeout_secs(1, sub_rx.next()).await, Err(_));
```

Comment:
@niklasad1 ^^^

Reply:
you need to drop the subscription for this to work, on master the subscription is dropped inside the block scope and is dropped when it reached here,